exporter/azuredataexplorerexporter/adx_exporter.go

```diff
@@ -0,0 +1,281 @@
+// limitations under the License.
+
+package azuredataexplorerexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/azuredataexplorerexporter"
+
+import (
+	"context"
+	"io"
+	"strings"
+
+	"github.com/Azure/azure-kusto-go/kusto"
+	"github.com/Azure/azure-kusto-go/kusto/data/errors"
+	"github.com/Azure/azure-kusto-go/kusto/ingest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	jsoniter "github.com/json-iterator/go"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+// adxDataProducer uses the ADX client to perform ingestion
+type adxDataProducer struct {
+	client        *kusto.Client       // client for logs , traces and metrics
+	ingestor      localIngestor       // ingestion for  logs, traces and metrics
+	ingestoptions []ingest.FileOption // Options for the ingestion
+	logger        *zap.Logger         // Loggers for tracing the flow
+}
+
+type localIngestor interface {
+	FromReader(ctx context.Context, reader io.Reader, options ...ingest.FileOption) (*ingest.Result, error)
+	Close() error
+}
+
+const nextline = "\n"
+
+const (
+	// Scope name
+	scopename = "scope.name"
+	// Scope version
+	scopeversion = "scope.version"
+)
+
+// given the full metrics , extract each metric , resource attributes and scope attributes. Individual metric mapping is sent on to metricdata mapping
+func (e *adxDataProducer) metricsDataPusher(ctx context.Context, metrics pmetric.Metrics) error {
+	var metricsBuffer []string
+	transformedadxmetrics, err := rawMetricsToAdxMetrics(ctx, metrics, e.logger)
+	if err != nil {
+		e.logger.Error("Error transforming metrics to ADX metric format.", zap.Error(err))
+		return err
+	}
+	// Since the transform succeeded ,  using the option for ingestion ingest the data into ADX
+	for _, tm := range transformedadxmetrics {
+		adxmetricjsonstring, err := jsoniter.MarshalToString(tm)
+		if err != nil {
+			e.logger.Error("Error performing serialization of data.", zap.Error(err))
+		}
+		metricsBuffer = append(metricsBuffer, adxmetricjsonstring)
+
+	}
+	if len(metricsBuffer) != 0 {
+		if err := e.ingestData(metricsBuffer); err != nil {
+			return err
+		}
+	}
+	metricsflushed := len(transformedadxmetrics)
+	e.logger.Sugar().Infof("Flushing %d metrics to sink", metricsflushed)
+	return nil
+}
+
+func (e *adxDataProducer) ingestData(b []string) error {
+
+	ingestreader := strings.NewReader(strings.Join(b, nextline))
+
+	if _, err := e.ingestor.FromReader(context.Background(), ingestreader, e.ingestoptions...); err != nil {
+		e.logger.Error("Error performing managed data ingestion.", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+func (e *adxDataProducer) logsDataPusher(ctx context.Context, logData plog.Logs) error {
+	resourcelogs := logData.ResourceLogs()
+	var logsBuffer []string
+	for i := 0; i < resourcelogs.Len(); i++ {
```

Comment:
here and below: can you use ranges rather then indexes

Reply:
@yogilad the `logData.ResourceLogs()`  returns a `ResourceLogsSlice`. Which is a struct with a slice field in it. Otel has provided APIs like `At()`,  `Len()`, `EnsureCapacity()` to use and process the slice. 
![image](https://user-images.githubusercontent.com/102726227/176665861-56f297e4-86dc-4d58-85c0-6cc8c661bc53.png)